### PR TITLE
fix(): prerender variables

### LINKED ISSUE #2585: Discrepencies between manual and trigger execution in variabe handling
### Explain the bug

Depending if the execution is run manually or via a trigger, the `vars.period` variable doesn't have the same behavior; especially on the `replace` Pebble method:

```yaml
id: copine_voyage_troubleshoot
namespace: dev

variables:
  period: "{{ schedule.date ?? execution.startDate | dateAdd(-1, 'DAYS') | date('yyyy-MM') }}"

tasks:

  - id: debug
    type: io.kestra.core.tasks.log.Log
    message: "Period: {{ vars.period }}, {{ vars.period | replace({'-': ''}) }}, {{ vars.period | replace({'2023': 'toto'}) }}"

triggers:
  - id: schedule
    type: io.kestra.core.models.triggers.types.Schedule
    cron: "0 12 1 * *"
    backfill:
      start: 2023-08-01T00:00:00Z
```

When running manually (working fine, except for the "toto" replacement):
image

When running from trigger (here backfill schedule):

image

### Environment Information

- Kestra Version: 0.13.0
- Operating System and Java Version (if not using Kestra Docker image):

### PR DESCRIPTION
**Pull request by Skraye**
close #2585

### CODE REVIEW

File: core/src/main/java/io/kestra/core/runners/RunContext.java [hunk 1]
```diff
@@ -434,12 +434,22 @@ public RunContext forWorkerDirectory(ApplicationContext applicationContext, Work
         return this;
     }
 
+    // Default behavior
     public String render(String inline) throws IllegalVariableEvaluationException {
-        return variableRenderer.render(inline, this.variables);
+        return variableRenderer.render(inline, this.variables, true);
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/RunContext.java:439]: `true` is the default so you can revert this change

File: core/src/main/java/io/kestra/core/runners/RunContext.java [hunk 2]
```diff
@@ -434,12 +434,22 @@ public RunContext forWorkerDirectory(ApplicationContext applicationContext, Work
...
+    }
+
+    public String render(String inline, boolean renderVariables) throws IllegalVariableEvaluationException {
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/RunContext.java:442]:
```suggestion
    public String render(String inline, boolean preprocessVariables) throws IllegalVariableEvaluationException {
```

I think it's better to call it preprocessed as it's what you do and variable are rendered anyway as we are recursive.

File: core/src/main/java/io/kestra/core/runners/RunContext.java [hunk 3]
```diff
@@ -434,12 +434,22 @@ public RunContext forWorkerDirectory(ApplicationContext applicationContext, Work
...
+        return variableRenderer.render(inline, this.variables, renderVariables);
     }
 
+    // Default behavior
     public String render(String inline, Map<String, Object> variables) throws IllegalVariableEvaluationException {
-        return variableRenderer.render(inline, mergeVariables(variables));
+        return variableRenderer.render(inline, mergeVariables(variables), true);
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/RunContext.java:448]: Revert as it's the default behaviour

File: core/src/main/java/io/kestra/core/runners/RunContext.java
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/RunContext.java:1]: I think all changes inside the RunContext can be reverted as the new functions are not used anywhere.
Just keep it simple for now

File: core/src/main/java/io/kestra/core/runners/VariableRenderer.java [hunk 4]
```diff
@@ -93,6 +93,10 @@ public VariableRenderer(ApplicationContext applicationContext, @Nullable Variabl
     }
 
     public String recursiveRender(String inline, Map<String, Object> variables) throws IllegalVariableEvaluationException {
+        return recursiveRender(inline, variables, true);
+    }
+
+    public String recursiveRender(String inline, Map<String, Object> variables, Boolean renderVariables) throws IllegalVariableEvaluationException {
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/VariableRenderer.java:99]:
```suggestion
    public String recursiveRender(String inline, Map<String, Object> variables, Boolean preprocessVariables) throws IllegalVariableEvaluationException {
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I didn't see any test with a recursive rendering, please create one, we may want to throw is a loop is detected instead of doing a bad rendering.

Also, it appears you change the default behaviour and allow to use the old behavior but there is not usage of the old so it seems complicated for nothing.

File: core/src/test/java/io/kestra/core/runners/RunContextTest.java [hunk 5]
```diff
@@ -112,14 +113,15 @@ void inputsLarge() throws TimeoutException {
...
-        assertThat(execution.getTaskRunList(), hasSize(3));
+        assertThat(execution.getTaskRunList(), hasSize(4));
 
         assertThat(
             ZonedDateTime.from(ZonedDateTime.parse((String) execution.getTaskRunList().get(0).getOutputs().get("value"))),
             ZonedDateTimeMatchers.within(10, ChronoUnit.SECONDS, ZonedDateTime.now())
         );
         assertThat(execution.getTaskRunList().get(1).getOutputs().get("value"), is("task-id"));
         assertThat(execution.getTaskRunList().get(2).getOutputs().get("value"), is("return"));
+        assertThat((String) execution.getTaskRunList().get(3).getOutputs().get("value"), containsString("toto"));
```
Comment (brian-mulier-p) [inline core/src/test/java/io/kestra/core/runners/RunContextTest.java:124]: Please add an infinite loop test and assert that you get the result that preceeds the infinite loop turn
(var1: {{var2}}, var2: {{var3}}, var3: {{var1}})
=> render({{var1 | replace({'3': '4'})}}) => {{var4}}

Comment (brian-mulier-p) [reply core/src/test/java/io/kestra/core/runners/RunContextTest.java:124]: you can change the test as it's purely logic and not easily understandable but you get it

**Review by brian-mulier-p — CHANGES_REQUESTED** [review]

**Comment by Skraye** [comment]
The `variables()` test has been updated

You can see that the task added to the flow depends on a vars, and the vars.period itself need to be rendered first, how the value return by the task would not contains any "toto"

```yaml
- id: variables
  type: io.kestra.core.tasks.debugs.Return
  format: "{{ vars.period | replace({':': 'toto'}) }}"
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I really think it is a bad idea to store all intermediate rendering in a list to detect infinite loop.
It could blow away the memory if a big string is rendered (and we now it is).

If it's still doable today to create an infinite loop, remove your infinite loop handling code and proceed with the PR and open an issue. If not, we should hold this one.

**Comment by Skraye** [comment]
@loicmathieu issue for the infinite loop : https://github.com/kestra-io/kestra/issues/2592
Pls approve if you consider the actual state is ok for variable rendering before rendering dynamic props

**Review by loicmathieu — APPROVED** [review]
I did't had time to QA it, @brian-mulier-p did you QA it?

**Comment by yuri1969** [comment]
FYI this PR also fixes #2466

**Review by anna-geller — APPROVED** [review]
QA pass

**Pull request merged**